akka24-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/BundlesConnector.scala

```diff
@@ -0,0 +1,206 @@
+
+  sealed trait Message
+
+  private case object ConsumeBundlesEvents extends Message
+  private case class ConsumeBundlesEventsOutcome(outcome: Try[Done]) extends Message
+
+  private case object GetBundles extends Message
+  private case class GetBundlesSuccess(bundles: Seq[Bundle]) extends Message
+  private case class GetBundlesError(error: Throwable) extends Message
+
+  case class StopInstruction(delay: FiniteDuration)(val callback: () => Unit) extends Message
+
+  sealed trait State
+
+  private object ExpectingEvent extends State
+  private object ExpectingEventBurst extends State
+
+  sealed trait CallbackResult
+  object CallbackResult {
+    case object Continue extends CallbackResult
+    case object Stop extends CallbackResult
+  }
+
+  def props[T](
+    getBundlesEvents: BundlesConnector.GetBundleEvents,
+    getBundlesInfo: BundlesConnector.GetBundlesInfo,
+    expectingEventBurstTimeout: FiniteDuration = 500.millis,
+    reconnectInterval: FiniteDuration = 500.millis,
+    stopInstruction: Option[StopInstruction] = None
+  )(callback: BundlesConnector.Callback): Props =
+    Props(new BundlesConnector(getBundlesEvents, getBundlesInfo, expectingEventBurstTimeout, reconnectInterval, stopInstruction)(callback))
+
+  private def getBundles(connector: ActorRef, getBundlesInfo: GetBundlesInfo)(implicit ec: ExecutionContext): Unit =
+    getBundlesInfo()
+      .map(GetBundlesSuccess)
+      .recover {
+        case e => GetBundlesError(e)
+      }
+      .pipeTo(connector)
+
+  /*
+   * When consuming bundle events we request the retrieval of bundle info only when a connection
+   * to the events stream is established. This ensures that we don't have a race condition whereby
+   * we miss events. Otherwise, if we retrieved the bundles before being connected to the event
+   * stream then we could miss out on events. We also cannot retrieve the bundle info only when
+   * having received events because then startup could fail if there are none i.e. we must always
+   * get the initial state of bundle info.
+   */
+  private def consumeBundlesEvents(connector: ActorRef, getBundlesEvents: BundlesConnector.GetBundleEvents)(
+    implicit
+    ec: ExecutionContext, mat: ActorMaterializer
+  ): Unit =
+    for {
+      eventStreamResult <- getBundlesEvents()
+    } eventStreamResult match {
+      case EventStreamSuccess(sseSource) =>
+        sseSource
+          .runForeach(connector ! _)
+          .map(done => connector ! ConsumeBundlesEventsOutcome(Success(done)))
+          .recover {
+            case error =>
+              connector ! ConsumeBundlesEventsOutcome(Failure(error))
+          }
+
+      case EventStreamFailure(httpResponseCode, errorMessage) =>
+        connector ! ConsumeBundlesEventsOutcome(
+          Failure(new Exception(s"HTTP failure when obtaining bundle events - code [$httpResponseCode] - error [$errorMessage]"))
+        )
+    }
+}
```

Comment:
I can't quite put my finger on it, but it feels like the `Streams.sseFlow` of akka-sse has been buried somewhere. I would have expected it to be within the `BundlesConnector`. Too many layers of indirection...